src/gevent/_gevent_cgreenlet.pxd

```diff
@@ -57,30 +57,61 @@ cdef extern from "Python.h":
     ctypedef class types.CodeType [object PyCodeObject]:
         pass
 
-cdef extern from "frameobject.h":
-
-    ctypedef class types.FrameType [object PyFrameObject]:
-        cdef CodeType f_code
-        # Accessing the f_lineno directly doesn't work. There is an accessor
-        # function, PyFrame_GetLineNumber that is needed to turn the raw line number
-        # into the executing line number.
-        # cdef int f_lineno
-        # We can't declare this in the object as an object, because it's
-        # allowed to be NULL, and Cython can't handle that.
-        # We have to go through the python machinery to get a
-        # proper None instead, or use an inline function.
-        cdef void* f_back
+IF PY311A6:
+    cdef extern from "internal/pycore_frame.h":
+        ctypedef class types._PyInterpreterFrame [object _PyInterpreterFrame]:
```

Comment:
I was thinking I could do something like:

```suggestion
        ctypedef struct _PyInterpreterFrame:
```

But that way, I get:

```
Error compiling Cython file:
------------------------------------------------------------
...
        pass

IF PY311A6:
    cdef extern from "internal/pycore_frame.h":
        ctypedef struct _PyInterpreterFrame:
            cdef CodeType f_code
           ^
------------------------------------------------------------

src/gevent/_gevent_cgreenlet.pxd:63:12: Expected an identifier, found 'cdef'
```

And when I remove `cdef` from `cdef CodeType f_code` as well, I get:

```
C struct/union member cannot be a Python object
```

Reply:
In if I cascade with using structs:

```
cdef extern from "Python.h":

    ctypedef struct PyCodeObject:
        pass

...

    cdef extern from "internal/pycore_frame.h":
        ctypedef struct _PyInterpreterFrame:
            PyCodeObject* f_code
```

I end up with:

```
cdef inline void set_f_code(FrameType frame, CodeType code):
    IF PY311A6:
        frame.f_frame.f_code = code
                              ^
------------------------------------------------------------

src/gevent/_gevent_cgreenlet.pxd:114:31: Cannot convert Python object to 'PyCodeObject *'
```

And I am getting into a seemingly neverending hole.

Reply:
This makes it compile:

```patch
diff --git a/src/gevent/_gevent_cgreenlet.pxd b/src/gevent/_gevent_cgreenlet.pxd
index 246773e..47669dc 100644
--- a/src/gevent/_gevent_cgreenlet.pxd
+++ b/src/gevent/_gevent_cgreenlet.pxd
@@ -57,13 +57,16 @@ cdef extern from "Python.h":
     ctypedef class types.CodeType [object PyCodeObject]:
         pass
 
+    ctypedef struct PyCodeObject:
+        pass
+
 IF PY311A6:
     cdef extern from "internal/pycore_frame.h":
-        ctypedef class types._PyInterpreterFrame [object _PyInterpreterFrame]:
-            cdef CodeType f_code
+        ctypedef struct _PyInterpreterFrame:
+            PyCodeObject* f_code
 
         ctypedef class types.FrameType [object PyFrameObject]:
-            cdef _PyInterpreterFrame f_frame
+            cdef _PyInterpreterFrame* f_frame
             # Accessing the f_lineno directly doesn't work. There is an accessor
             # function, PyFrame_GetLineNumber that is needed to turn the raw line number
             # into the executing line number.
@@ -108,7 +111,7 @@ cdef inline CodeType get_f_code(FrameType frame):
 
 cdef inline void set_f_code(FrameType frame, CodeType code):
     IF PY311A6:
-        frame.f_frame.f_code = code
+        frame.f_frame.f_code = <PyCodeObject *>code
     ELSE:
         frame.f_code = code
 ```

But it fails on rutime with:

```
    File "src/gevent/greenlet.py", line 689, in gevent._gevent_cgreenlet.Greenlet.spawn
      g = cls(*args, **kwargs)
    File "src/gevent/greenlet.py", line 301, in gevent._gevent_cgreenlet.Greenlet.__init__
      self.spawning_stack = _extract_stack(self.spawning_stack_limit)
    File "src/gevent/greenlet.py", line 164, in gevent._gevent_cgreenlet._extract_stack
      set_f_lineno(older_Frame.f_lineno, get_f_lineno(frame)) # pylint:disable=undefined-variable
  TypeError: Cannot convert int to frame
```